### PR TITLE
Implement Proximity Chat

### PR DESCRIPTION
**Pull request by MahemaS27**
PR for Implementing Proximity Chat in the backend:
- added logic in `CoveyTownController` to handle sending proximity messages to nearby listeners
- edited method in `TestUtils` to use a `Player` instead of ID to grab the location easier
- added tests for new logic
- removes the direct message setting in `TestUtils` method -> will save this for the DM PR

### CODE REVIEW

File: services/roomService/src/client/TestUtils.ts [hunk 1]
```diff
@@ -94,17 +95,28 @@ export function setSessionTokenAndTownID(coveyTownID: string, sessionToken: stri
...
+  if (type == MessageType.ProximityMessage){
+    return {
+      userName: nanoid(),
+      userId: player1.id,
+      location: player1.location,
+      messageContent: "Omg I'm a test",
+      timestamp,
+      type,
+      directMessageId: directMessageID,
+    };
```
Comment (MahemaS27) [inline services/roomService/src/client/TestUtils.ts:111]: @abigailhodge this is what I was talking about

File: services/roomService/src/client/TestUtils.ts [hunk 2]
```diff
@@ -94,17 +95,28 @@ export function setSessionTokenAndTownID(coveyTownID: string, sessionToken: stri
...
+  }
+  if (type == MessageType.ProximityMessage){
+    return {
+      userName: nanoid(),
+      userId: player1.id,
+      location: player1.location,
+      messageContent: "Omg I'm a test",
+      timestamp,
+      type,
+      directMessageId: directMessageID,
```
Comment (michelle-pine) [inline services/roomService/src/client/TestUtils.ts]: we shouldn't be setting the `directMessageId` unless it's a direct message, best to leave it out until we hit that case. we don't even need to set the `let` above unless it's a dm

File: services/roomService/src/lib/CoveyTownController.test.ts [hunk 3]
```diff
@@ -92,11 +96,46 @@ describe('CoveyTownController', () => {
...
+      proximityMockListeners.forEach(listener => testingTown.addTownListener(listener));
+      // update the player location so that they are way out of proximity
+      const newLocation: UserLocation = {x:1000, y:1000, rotation:'back', moving: true}
+      player.updateLocation(newLocation);
+      const message = TestUtils.createMessageForTesting(MessageType.ProximityMessage, player);
+
+      testingTown.receiveMessage(message);
+      proximityMockListeners.forEach(listener => expect(listener.onMessageReceived).not.toBeCalledWith(message));
+    });
+
```
Comment (michelle-pine) [inline services/roomService/src/lib/CoveyTownController.test.ts:142]: can we have a test where some players are notified but some are not?

File: services/roomService/src/lib/CoveyTownController.ts [hunk 4]
```diff
@@ -172,4 +176,18 @@ export default class CoveyTownController {
   disconnectAllPlayers(): void {
     this._listeners.forEach((listener) => listener.onTownDestroyed());
   }
+
+  calculateNearbyListeners(players: Player[], currentLocation: UserLocation): CoveyTownListener[] {
+    const isWithinCallRadius = (p: Player, location: UserLocation) => {
+      if (p.location && location) {
+        const dx = p.location.x - location.x;
+        const dy = p.location.y - location.y;
+        const d = Math.sqrt(dx * dx + dy * dy);
```
Comment (michelle-pine) [inline services/roomService/src/lib/CoveyTownController.ts:185]: ah geometry...how i haven't missed you

File: services/roomService/src/client/TestUtils.ts (hunk 2 above)
Comment (MahemaS27) [reply services/roomService/src/client/TestUtils.ts]: sounds good! I'll save it for the dm PR

**Review by michelle-pine — APPROVED** [review]
💃💃

**Review by anikarabenhorst — APPROVED** [review]
looks good :)

**Pull request merged**